testapi.pm

```diff
@@ -1252,25 +1257,27 @@ sub type_string {
         return;
     }
 
-    my $max_interval = $args{max_interval}       // 250;
-    my $wait         = $args{wait_screen_change} // 0;
-    my $wait_still   = $args{wait_still_screen}  // 0;
-    bmwqemu::log_call(string => $log, max_interval => $max_interval, wait_screen_changes => $wait, wait_still_screen => $wait_still);
+    my $max_interval   = $args{max_interval}       // 250;
+    my $wait           = $args{wait_screen_change} // 0;
+    my $wait_still     = $args{wait_still_screen}  // 0;
+    my $wait_timeout   = $args{timeout}            // 30;
+    my $wait_sim_level = $args{similarity_level}   // 47;
+    bmwqemu::log_call(string => $log, max_interval => $max_interval, wait_screen_changes => $wait, wait_still_screen => $wait_still,
+        timeout => $wait_timeout, similarity_level => $wait_sim_level);
+    my @pieces;
     if ($wait) {
         # split string into an array of pieces of specified size
         # https://stackoverflow.com/questions/372370
-        my @pieces = unpack("(a${wait})*", $string);
-        for my $piece (@pieces) {
-            wait_screen_change { query_isotovideo('backend_type_string', {text => $piece, max_interval => $max_interval}); };
-            if ($wait_still) {
-                wait_still_screen($wait_still);
-            }
-        }
+        @pieces = unpack("(a${wait})*", $string);
     }
     else {
-        query_isotovideo('backend_type_string', {text => $string, max_interval => $max_interval});
-        if ($wait_still) {
-            wait_still_screen($wait_still);
+        push @pieces, $string;
+    }
+    for my $piece (@pieces) {
+        wait_screen_change { query_isotovideo('backend_type_string', {text => $piece, max_interval => $max_interval}); };
```

Comment:
I think we have a problem here. In http://lord.arch.suse.de/tests/2443/file/autoinst-log.txt looking for `wait_screen_change` I found that `wait_screen_change` is called but I never requested it. The problem is that we always loop over `@pieces` now and call `wait_screen_change` even when not asked. In most cases this will not fail but just waste time when there is no actual screen change visible. In some cases this might fail in obscure ways, e.g. when no graphical console is actually connected, e.g. when typing `reboot` on a remote backend where the graphical session dies as soon as the command is effectively rebooting the SUT

Reply:
reported as https://progress.opensuse.org/issues/54935

Reply:
fix in https://github.com/os-autoinst/os-autoinst/pull/1183